### ISSUE TITLE
PagedList getTotalRowCount() includes soft deleted rows in its count

### ISSUE THREAD

**Issue by elennick** [ebean-orm/avaje-ebeanorm#511]
When getting the total row count using ```getTotalRowCount()``` from a ```PagedList```, the count that comes back includes soft deleted rows whereas the actual items coming back via the ```PagedList.getList()``` do not. This leads to an inaccurate total when you have soft deleted rows in the table.

Would it be possible to update getTotalRowCount() to not include soft deleted rows in its count? Or to create another method on PagedList called ```getTotalRowCountMinusSoftDeleted()``` or something to that effect?

**Comment by simongodard** [comment]
Hi @rbygrave, I still see the bug in the 6.17.3 release. `pagedList.getTotalRowCount()` includes soft-deleted beans in the count. The actual beans returned by `pagedList.getList()` don't include soft-deleted.

**Comment by rbygrave** [comment]
I have just pushed up a test that is passing (so this looks to be working correctly).  Can you review this test I have just added.

The transaction logging and asserts show that the soft delete predicate is included for the total row count query.

```sql
select count(*) from cover t0 where t0.s3url like ? and coalesce(t0.deleted,false)=false; --bind(SoftDelPaged-%)
```

The log for this test is:

```console

08:15:50.305 [main] DEBUG org.avaje.ebean.SUM - txn[1004] BatchControl flush [Cover:100 i:3]
08:15:50.308 [main] DEBUG org.avaje.ebean.SQL - txn[1004] insert into cover (id, s3url, deleted) values (?,?,?); --bind(1,SoftDelPaged-1,false,)
08:15:50.308 [main] DEBUG org.avaje.ebean.SQL - txn[1004] insert into cover (id, s3url, deleted) values (?,?,?); --bind(2,SoftDelPaged-2,false,)
08:15:50.309 [main] DEBUG org.avaje.ebean.SQL - txn[1004] insert into cover (id, s3url, deleted) values (?,?,?); --bind(3,SoftDelPaged-3,false,)
08:15:50.309 [main] DEBUG org.avaje.ebean.SUM - txn[1004] Inserted [Cover] [1]
08:15:50.309 [main] DEBUG org.avaje.ebean.SUM - txn[1004] Inserted [Cover] [2]
08:15:50.309 [main] DEBUG org.avaje.ebean.SUM - txn[1004] Inserted [Cover] [3]
08:15:50.310 [main] DEBUG org.avaje.ebean.TXN - txn[1004] Commit
08:15:50.311 [main] DEBUG org.avaje.ebean.SQL - txn[1005] update cover set deleted=? where id=?; --bind(true,2,)
08:15:50.311 [main] DEBUG org.avaje.ebean.SUM - txn[1005] SoftDelete [Cover] [2]
08:15:50.311 [main] DEBUG org.avaje.ebean.TXN - txn[1005] Commit
08:15:50.332 [main] DEBUG org.avaje.ebean.SQL - txn[1006] select count(*) from cover t0 where t0.s3url like ? and coalesce(t0.deleted,false)=false; --bind(SoftDelPaged-%)
08:15:50.332 [main] DEBUG org.avaje.ebean.SUM - txn[1006] FindRowCount exeMicros[1872] rows[2] type[com.avaje.tests.model.onetoone.album.Cover] predicates[t0.s3url like ? ] bind[SoftDelPaged-%]
08:15:50.333 [main] TRACE org.avaje.ebean.TXN - txn[1006] Commit - query only
08:15:50.337 [main] DEBUG org.avaje.ebean.SQL - txn[1007] select t0.id c0, t0.s3url c1, t0.deleted c2 from cover t0 where t0.s3url like ? and coalesce(t0.deleted,false)=false order by t0.id limit 10; --bind(SoftDelPaged-%)
08:15:50.338 [main] DEBUG org.avaje.ebean.SUM - txn[1007] FindMany type[Cover] origin[Ba7fGt.CTq5MT.BvQ020] exeMicros[2139] rows[2] name[] predicates[t0.s3url like ? ] bind[SoftDelPaged-%]
08:15:50.339 [main] TRACE org.avaje.ebean.TXN - txn[1007] Commit - query only
```

And the code for this test is:

```java
    List<Cover> list = new ArrayList<Cover>();
    list.add(new Cover("SoftDelPaged-1"));
    list.add(new Cover("SoftDelPaged-2"));
    list.add(new Cover("SoftDelPaged-3"));

    Ebean.saveAll(list);
    Ebean.delete(list.get(1));

    LoggedSqlCollector.start();

    PagedList<Cover> pagedList = Ebean.find(Cover.class)
        .where().startsWith("s3Url", "SoftDelPaged-")
        .setMaxRows(10)
        .findPagedList();

    int totalRowCount = pagedList.getTotalRowCount();
    List<Cover> resultList = pagedList.getList();

    List<String> sql = LoggedSqlCollector.stop();

    assertThat(resultList).hasSize(2);
    assertThat(totalRowCount).isEqualTo(2);

    assertThat(sql).hasSize(2);
    assertThat(sql.get(0)).contains("select count(*) from cover t0 where t0.s3url like");
    assertThat(sql.get(0)).contains("and coalesce(t0.deleted,false)=false; --bind(SoftDelPaged-%)");

    assertThat(sql.get(1)).contains("where t0.s3url like ");
    assertThat(sql.get(1)).contains("and coalesce(t0.deleted,false)=false order by t0.id");

```

**Comment by simongodard** [comment]
Hi @rbygrave, your test proved without a doubt that something was wrong on my end. I just realized this morning that even if I add a Maven dependency on the latest _avaje-ebeanorm_ version 7.11.2, the effective dependency was overriden by a transitive dependency from _avaje-ebeanorm-spring_ and I was ending up with version 6.13.3 (before the fix which was included in 6.14.3).

I'm sorry about that! The good thing is we have this much clearer test now to prove the fix.

I just tested with the correct version and everything is fine.

Best regards,
Simon

**Comment by rbygrave** [comment]
>  this much clearer test now to prove the fix.
Yes. The prior test was a bit of a shocker really.

> something was wrong on my end.
Great to get that squared away.

> was overridden by a transitive dependency from avaje-ebeanorm-spring
Ok. There is a good case for changing that dependency in avaje-ebeanorm-spring to be `provided` scope to avoid this issue in the future - I think I'll make that change to avaje-ebeanorm-spring.

Thanks, Rob.
